### PR TITLE
JAMES-3295 matcher temporary delivery error

### PR DESCRIPTION
**Pull request by Arsnael**
Following up Lan's work on #3562

### CODE REVIEW

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/RemoteDeliveryErrorHandlingTest.java [hunk 1]
```diff
@@ -108,7 +109,7 @@ void setup(DockerMockSmtp dockerMockSmtp) throws Exception {
                         .matcher(IsRemoteDeliveryPermanentError.class)
                         .addProperty("repositoryPath", REMOTE_DELIVERY_PERMANENT_ERROR_REPOSITORY.asString()))
                     .addMailet(MailetConfiguration.builder()
-                        .matcher(All.class)
+                        .matcher(IsRemoteDeliveryTemporaryError.class)
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/RemoteDeliveryErrorHandlingTest.java:112]: I don't think that is needed. proof: you re-enabled no tests.

Solution: 

Manage to have a mail getting in the RemoteDelivery but reaching this "by accident", in which case we don't want to store it as a temporary error but maybe in another dedicated mail repository.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryTemporaryError.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.james.core.MailAddress;
+import org.apache.mailet.AttributeUtils;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMatcher;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * <p>
+ * Checks if the mail has a permanent remote delivery failure attribute set to false (meaning it's a temporary error)
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/IsRemoteDeliveryTemporaryError.java:36]: Checks if the mail has a temporary remote delivery failure

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/RemoteDeliveryErrorHandlingTest.java [hunk 3]
```diff
@@ -242,4 +254,18 @@ void remoteDeliveryShouldStoreTemporaryFailureAsPermanentWhenExceedsMaximumRetri
                 .getRepositoryMailCount(REMOTE_DELIVERY_PERMANENT_ERROR_REPOSITORY))
                 .isEqualTo(1));
     }
+
+    @Test
+    void remoteDeliveryShouldStoreErrorNotPermanentAndNotTemporarySeparately(SMTPMessageSender smtpMessageSender) throws Exception {
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/RemoteDeliveryErrorHandlingTest.java:259]: remoteDeliveryErrorHandlingShouldIgnoreMailsNotTransitingByRemoteDelivery

**Comment by Arsnael** [comment]
Rewrote the history, as the test name is part of the commit description

**Comment by Arsnael** [comment]
Rebased on master, but checking the failed build right before, saw this:

```
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR] Failures: 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR]   DeliveryRunnableTest.deliveryTemporaryFailureShouldResetDeliveryCountOnNonErrorState:238 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Argument(s) are different! Wanted:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=1, serializer=IntSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT1S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnableTest.deliveryTemporaryFailureShouldResetDeliveryCountOnNonErrorState(DeliveryRunnableTest.java:238)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Actual invocations have different arguments:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=1, serializer=IntSerializer}}, AttributeName{name=is-delivery-permanent-error}=Attribute{name=AttributeName{name=is-delivery-permanent-error}, value=AttributeValue{value=false, serializer=BooleanSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT1S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnable.reAttemptDelivery(DeliveryRunnable.java:175)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR]   DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDelivery:141 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Argument(s) are different! Wanted:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=1, serializer=IntSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT1S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDelivery(DeliveryRunnableTest.java:141)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Actual invocations have different arguments:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=1, serializer=IntSerializer}}, AttributeName{name=is-delivery-permanent-error}=Attribute{name=AttributeName{name=is-delivery-permanent-error}, value=AttributeValue{value=false, serializer=BooleanSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT1S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnable.reAttemptDelivery(DeliveryRunnable.java:175)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR]   DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDeliveryOnMaximumRetryNumber:185 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Argument(s) are different! Wanted:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=5, serializer=IntSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT5S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDeliveryOnMaximumRetryNumber(DeliveryRunnableTest.java:185)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Actual invocations have different arguments:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=5, serializer=IntSerializer}}, AttributeName{name=is-delivery-permanent-error}=Attribute{name=AttributeName{name=is-delivery-permanent-error}, value=AttributeValue{value=false, serializer=BooleanSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT5S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnable.reAttemptDelivery(DeliveryRunnable.java:175)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR]   DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDeliveryWithRightDelay:163 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Argument(s) are different! Wanted:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=3, serializer=IntSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT3S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDeliveryWithRightDelay(DeliveryRunnableTest.java:163)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Actual invocations have different arguments:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=3, serializer=IntSerializer}}, AttributeName{name=is-delivery-permanent-error}=Attribute{name=AttributeName{name=is-delivery-permanent-error}, value=AttributeValue{value=false, serializer=BooleanSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT3S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnable.reAttemptDelivery(DeliveryRunnable.java:175)
```
It's related. Fixing...

**Comment by chibenwa** [comment]
```
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR] Failures: 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR]   DeliveryRunnableTest.deliveryTemporaryFailureShouldResetDeliveryCountOnNonErrorState:238 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Argument(s) are different! Wanted:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=1, serializer=IntSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT1S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnableTest.deliveryTemporaryFailureShouldResetDeliveryCountOnNonErrorState(DeliveryRunnableTest.java:238)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Actual invocations have different arguments:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=1, serializer=IntSerializer}}, AttributeName{name=is-delivery-permanent-error}=Attribute{name=AttributeName{name=is-delivery-permanent-error}, value=AttributeValue{value=false, serializer=BooleanSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT1S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnable.reAttemptDelivery(DeliveryRunnable.java:175)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR]   DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDelivery:141 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Argument(s) are different! Wanted:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=1, serializer=IntSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT1S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDelivery(DeliveryRunnableTest.java:141)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Actual invocations have different arguments:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=1, serializer=IntSerializer}}, AttributeName{name=is-delivery-permanent-error}=Attribute{name=AttributeName{name=is-delivery-permanent-error}, value=AttributeValue{value=false, serializer=BooleanSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT1S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnable.reAttemptDelivery(DeliveryRunnable.java:175)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR]   DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDeliveryOnMaximumRetryNumber:185 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Argument(s) are different! Wanted:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=5, serializer=IntSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT5S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDeliveryOnMaximumRetryNumber(DeliveryRunnableTest.java:185)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Actual invocations have different arguments:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=5, serializer=IntSerializer}}, AttributeName{name=is-delivery-permanent-error}=Attribute{name=AttributeName{name=is-delivery-permanent-error}, value=AttributeValue{value=false, serializer=BooleanSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT5S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnable.reAttemptDelivery(DeliveryRunnable.java:175)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] [ERROR]   DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDeliveryWithRightDelay:163 
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Argument(s) are different! Wanted:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=3, serializer=IntSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT3S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnableTest.deliveryTemporaryFailureShouldRetryDeliveryWithRightDelay(DeliveryRunnableTest.java:163)
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] Actual invocations have different arguments:
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] mailQueue.enQueue(
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     FakeMail{msg=null, recipients=[], name=name, sender=null, state=error, errorMessage=null, lastUpdated=Sat Sep 30 06:53:14 GMT 2006, attributes={AttributeName{name=delivery_retry_count}=Attribute{name=AttributeName{name=delivery_retry_count}, value=AttributeValue{value=3, serializer=IntSerializer}}, AttributeName{name=is-delivery-permanent-error}=Attribute{name=AttributeName{name=is-delivery-permanent-error}, value=AttributeValue{value=false, serializer=BooleanSerializer}}}, size=0, remoteAddr=127.0.0.1},
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a]     PT3S
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] );
[ea243945ddc2bd6c667a829a16a3528f7a4ca80a] -> at org.apache.james.transport.mailets.remote.delivery.DeliveryRunnable.reAttemptDelivery(DeliveryRunnable.java:175)
```

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**